paywall/src/data-iframe/blockchainHandler/purchaseKey/submittedListener.js

```diff
@@ -0,0 +1,83 @@
+import { getAccount } from '../account'
+import { linkTransactionsToKeys } from '../keyStatus'
+import { getNetwork } from '../network'
+
+/**
+ * Listen for a submitted key purchase transaction
+ *
+ * If the latest transaction for a key is already submitted or newer, we immediately return
+ * in order to allow declarative chaining of transaction listeners.
+ *
+ * This allows us to call the transaction listener sequence for all transactions, including when
+ * the user refreshes the page in the middle of a transaction sequence. In this case, the
+ * submitted listener will skip its work and pass to the next one in the sequence
+ *
+ * @param {string} lockAddress the address of the lock we are listening for a submitted key purchase transaction
+ * @param {object} existingTransactions transactions, indexed by hash
+ * @param {object} existingKeys keys, indexed by "lock address-owner address"
+ * @param {walletService} walletService the walletService that initiated the key transaction
+ * @param {int} requiredConfirmations the number of required confirmations for a transaction to be considered permanent
+ */
+export default async function submittedListener({
+  lockAddress,
+  existingTransactions,
+  existingKeys,
+  walletService,
+  requiredConfirmations,
+}) {
+  // update key status for expired keys
+  const keys = linkTransactionsToKeys({
+    keys: existingKeys,
+    transactions: existingTransactions,
+    locks: [lockAddress],
+    requiredConfirmations,
+  })
+  const account = getAccount()
+  const network = getNetwork()
+  const keyToPurchase = `${lockAddress}-${account}`
+  const key = keys[keyToPurchase]
+  if (
+    (key.status !== 'none' && key.status !== 'expired') ||
+    key.status === 'failed'
```

Comment:
I really think we should constant-ize the key status here because I have to say I am very blurry on what are the "allowed" values for a status. What does 'failed' mean actually? Is that the transaction status? If so is that the latest transaction or another one?

Reply:
The key status is always linked to the most recent transaction.

Reply:
You can see how the key status is generated in `keyStatus.js` which is in master

Reply:
here: https://github.com/unlock-protocol/unlock/blob/master/paywall/src/data-iframe/blockchainHandler/keyStatus.js#L19

Reply:
I pushed a commit, it doesn't "constant-ize" the statuses, as we don't have direct control over some of the statuses that come from transaction status. It does fix the logic inconsistency: we should allow purchasing a new key when the previous transaction failed.